File: ios/Runner/SwiftEngine/EchoMessage.swift
import Foundation

/// Message types flowing through the interpretation pipeline.
/// Serialized as dictionaries and sent to Flutter via EventChannel.
enum EchoMessageType: Int {
    case asrPartial = 0
    case asrConfirmed = 1
    case translationStream = 2
    case translationDone = 3
    case error = 4
    case thermalState = 5
    case latencyWarning = 6
    case engineReady = 7
}

/// A typed message that can be sent to Flutter.
struct EchoMessage {
    let type: EchoMessageType
    let speakerId: Int
    let text: String
    let segmentId: Int
    let timestampMs: Int64
    let errorCode: Int
    let detail: String

    /// Convert to a dictionary for FlutterEventSink.
    func toMap() -> [String: Any] {
        return [
            "type": type.rawValue,
            "speakerId": speakerId,
            "text": text,
            "segmentId": segmentId,
            "timestampMs": timestampMs,
            "errorCode": errorCode,
            "detail": detail,
        ]
    }

    // Convenience constructors
    static func asrPartial(speakerId: Int, text: String, segmentId: Int) -> EchoMessage {
        EchoMessage(type: .asrPartial, speakerId: speakerId, text: text,
                    segmentId: segmentId, timestampMs: Int64(Date().timeIntervalSince1970 * 1000),
                    errorCode: 0, detail: "")
    }

    static func asrConfirmed(speakerId: Int, text: String, segmentId: Int) -> EchoMessage {
        EchoMessage(type: .asrConfirmed, speakerId: speakerId, text: text,
                    segmentId: segmentId, timestampMs: Int64(Date().timeIntervalSince1970 * 1000),
                    errorCode: 0, detail: "")
    }

    static func translationStream(speakerId: Int, token: String, segmentId: Int) -> EchoMessage {
        EchoMessage(type: .translationStream, speakerId: speakerId, text: token,
                    segmentId: segmentId, timestampMs: Int64(Date().timeIntervalSince1970 * 1000),
                    errorCode: 0, detail: "")
    }

    static func translationDone(speakerId: Int, text: String, segmentId: Int) -> EchoMessage {
        EchoMessage(type: .translationDone, speakerId: speakerId, text: text,
                    segmentId: segmentId, timestampMs: Int64(Date().timeIntervalSince1970 * 1000),
                    errorCode: 0, detail: "")
    }

    static func error(code: Int, detail: String) -> EchoMessage {
        EchoMessage(type: .error, speakerId: 0, text: "", segmentId: 0,
                    timestampMs: Int64(Date().timeIntervalSince1970 * 1000),
                    errorCode: code, detail: detail)
    }

    static func engineReady(status: String) -> EchoMessage {
        EchoMessage(type: .engineReady, speakerId: 0, text: status, segmentId: 0,
                    timestampMs: Int64(Date().timeIntervalSince1970 * 1000),
                    errorCode: 0, detail: "")
    }

    static func thermalState(mode: Int, detail: String) -> EchoMessage {
        EchoMessage(type: .thermalState, speakerId: 0, text: "", segmentId: 0,
                    timestampMs: Int64(Date().timeIntervalSince1970 * 1000),
                    errorCode: mode, detail: detail)
    }
}

File: android/app/src/main/kotlin/com/example/qwen_echo/engine/EchoMessage.kt
package com.example.qwen_echo.engine

import android.util.Log

/// Message types flowing through the interpretation pipeline.
/// Serialized as dictionaries and sent to Flutter via EventChannel.
/// Must match iOS EchoMessageType raw values exactly.
enum class EchoMessageType(val value: Int) {
    ASR_PARTIAL(0),
    ASR_CONFIRMED(1),
    TRANSLATION_STREAM(2),
    TRANSLATION_DONE(3),
    ERROR(4),
    THERMAL_STATE(5),
    LATENCY_WARNING(6),
    ENGINE_READY(7);
}

/// A locked audio segment ready for ASR processing.
data class LockedSegment(
    val audioData: ShortArray,
    val segmentId: Int,
    val speakerId: Int,
    val timestampMs: Long,
)

/// A typed message that can be sent to Flutter via EventChannel.
/// toMap() output must match iOS EchoMessage.toMap() exactly.
data class EchoMessage(
    val type: EchoMessageType,
    val speakerId: Int = 0,
    val text: String = "",
    val segmentId: Int = 0,
    val timestampMs: Long = System.currentTimeMillis(),
    val errorCode: Int = 0,
    val detail: String = "",
) {
    /// Convert to a dictionary for FlutterEventSink.
    /// Keys and types must match iOS exactly.
    fun toMap(): Map<String, Any> = mapOf(
        "type" to type.value,
        "speakerId" to speakerId,
        "text" to text,
        "segmentId" to segmentId,
        "timestampMs" to timestampMs,
        "errorCode" to errorCode,
        "detail" to detail,
    )

    companion object {
        private const val TAG = "EchoMessage"

        fun asrPartial(speakerId: Int, text: String, segmentId: Int) = EchoMessage(
            type = EchoMessageType.ASR_PARTIAL,
            speakerId = speakerId,
            text = text,
            segmentId = segmentId,
        )

        fun asrConfirmed(speakerId: Int, text: String, segmentId: Int) = EchoMessage(
            type = EchoMessageType.ASR_CONFIRMED,
            speakerId = speakerId,
            text = text,
            segmentId = segmentId,
        )

        fun translationStream(speakerId: Int, token: String, segmentId: Int) = EchoMessage(
            type = EchoMessageType.TRANSLATION_STREAM,
            speakerId = speakerId,
            text = token,
            segmentId = segmentId,
        )

        fun translationDone(speakerId: Int, text: String, segmentId: Int) = EchoMessage(
            type = EchoMessageType.TRANSLATION_DONE,
            speakerId = speakerId,
            text = text,
            segmentId = segmentId,
        )

        fun error(code: Int, detail: String) = EchoMessage(
            type = EchoMessageType.ERROR,
            errorCode = code,
            detail = detail,
        )

        fun engineReady(status: String) = EchoMessage(
            type = EchoMessageType.ENGINE_READY,
            text = status,
        )

        fun thermalState(mode: Int, detail: String) = EchoMessage(
            type = EchoMessageType.THERMAL_STATE,
            errorCode = mode,
            detail = detail,
        )
    }
}
